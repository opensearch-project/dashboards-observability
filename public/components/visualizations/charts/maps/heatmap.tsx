/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useMemo } from 'react';
import { colorPalette } from '@elastic/eui';
import { forEach, isEmpty, map } from 'lodash';
import Plotly from 'plotly.js-dist';
import {
  HEATMAP_PALETTE_COLOR,
  HEATMAP_SINGLE_COLOR,
  OPACITY,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';
import {
  hexToRgb,
  lightenColor,
  getPropName,
} from '../../../../components/event_analytics/utils/utils';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { Plt } from '../../plotly/plot';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import { PLOT_MARGIN } from '../../../../../common/constants/shared';
import {
  getCompleteTimespanKey,
  removeBackTick,
} from '../../../visualizations/charts/shared/common';
import { removeBacktick } from '../../../../../common/utils';

export const HeatMap = ({ visualizations, layout, config }: any) => {
  const {
    data: {
      explorer: {
        explorerData: { jsonData: fieldValueMapList },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          span = {},
          legend = {},
          tooltipOptions = {},
          panelOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        } = {},
        layoutConfig = {},
      } = {},
    } = {},
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const combinedDemensions = [
    ...(!isEmpty(span) ? [getCompleteTimespanKey(span)] : []),
    ...dimensions,
  ];

  const xaxisField = { ...combinedDemensions[0] };
  const yaxisField = { ...combinedDemensions[1] };
  const zMetrics = { ...series[0] };
  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';

  const colorField = chartStyles
    ? chartStyles.colorMode && chartStyles.colorMode[0].name === OPACITY
      ? chartStyles.color ?? HEATMAP_SINGLE_COLOR
      : chartStyles.scheme ?? HEATMAP_PALETTE_COLOR
    : HEATMAP_PALETTE_COLOR;
  const showColorscale = legend.showLegend ?? 'show';

  const traceColor: any = [];
  if (colorField.name === SINGLE_COLOR_PALETTE) {
    const colorsArray = colorPalette([lightenColor(colorField.color, 50), colorField.color], 10);
    colorsArray.map((hexCode, index) => {
      traceColor.push([
        (index !== colorsArray.length - 1 ? index : 10) / 10,
        hexToRgb(hexCode, 1, false),
      ]);
    });
  }

  const heatMapAxes = useMemo(() => {
    const dmaps = new Map<string, any>(); // key: values of 1st and 2nd group-by fields combined
    const uniqueXaxisVals = new Set<any>(); // any value for a field
    const uniqueYaxisVals = new Set<any>(); // any value for a field
    const zKey = getPropName(zMetrics);

    forEach(fieldValueMapList, (entry) => {
      const backtickRemovedEntry = removeBackTick(entry);
      const xKey = removeBacktick(xaxisField.label);
      const yKey = removeBacktick(yaxisField.label);

      // collect unique values from all values of 1st and 2nd group-by fields
      // for later composing 2 dimensional heatmap x, y axes
      uniqueXaxisVals.add(backtickRemovedEntry[xKey]);
      uniqueYaxisVals.add(backtickRemovedEntry[yKey]);

      // establish 1st,2nd -> data entry mapping for later filling in
      // corresponding aggregations to 2 dimensional heatmap zaxis
      dmaps.set(
        `${backtickRemovedEntry[xKey]},${backtickRemovedEntry[yKey]}`,
        backtickRemovedEntry
      );
    });

    const xAxis = [...uniqueXaxisVals];
    const yAxis = [...uniqueYaxisVals];
    const zaxis = map(yAxis, (yvalue) => {
      return map(xAxis, (xvalue) => {
        return dmaps.get(`${xvalue},${yvalue}`) ? dmaps.get(`${xvalue},${yvalue}`)[zKey] : null;
      });
    });

    return {
      z: zaxis,
      x: xAxis,
      y: yAxis,
    };
  }, [fieldValueMapList]);

  const heapMapData: Plotly.Data[] = [
    {
      ...heatMapAxes,
      hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
      colorscale: colorField.name === SINGLE_COLOR_PALETTE ? traceColor : colorField.name,
      type: 'heatmap',
      showscale: showColorscale === 'show',
    },
  ];

  layout.yaxis = { autosize: true, automargin: true };
  const mergedLayout = {
    ...layout,
    ...(layoutConfig.layout && layoutConfig.layout),
    title: panelOptions.title || layoutConfig.layout?.title || '',
    margin: PLOT_MARGIN,
  };

  const mergedConfigs = useMemo(
    () => ({
      ...config,
      ...(layoutConfig.config && layoutConfig.config),
    }),
    [config, layoutConfig.config]
  );

  return <Plt data={heapMapData} layout={mergedLayout} config={mergedConfigs} />;
};
