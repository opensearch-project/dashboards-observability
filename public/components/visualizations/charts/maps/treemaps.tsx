/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useMemo } from 'react';
import { indexOf, isEmpty, isEqual, isNull, uniq } from 'lodash';

import { Plt } from '../../plotly/plot';
import { EmptyPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/empty_placeholder';
import { NUMERICAL_FIELDS } from '../../../../../common/constants/shared';
import {
  DEFAULT_PALETTE,
  MULTI_COLOR_PALETTE,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';

export const TreeMap = ({ visualizations, layout, config }: any) => {
  const {
    data,
    metadata: { fields },
  } = visualizations.data.rawVizData;
  const { dataConfig = {}, layoutConfig = {} } = visualizations?.data?.userConfigs;

  const childField =
    dataConfig?.valueOptions &&
    dataConfig?.valueOptions.childField &&
    !isEmpty(dataConfig?.valueOptions.childField)
      ? dataConfig?.valueOptions.childField[0]
      : fields[fields.length - 1];

  const parentField =
    dataConfig?.valueOptions &&
    dataConfig?.valueOptions.parentField &&
    !isEmpty(dataConfig?.valueOptions.parentField)
      ? dataConfig?.valueOptions.parentField[0]
      : null;

  const valueField =
    dataConfig?.valueOptions &&
    dataConfig?.valueOptions.valueField &&
    !isEmpty(dataConfig?.valueOptions.valueField)
      ? dataConfig?.valueOptions.valueField[0]
      : fields[0];

  const colorField =
    dataConfig?.chartStyles && dataConfig?.chartStyles.colorTheme
      ? dataConfig?.chartStyles.colorTheme
      : { name: DEFAULT_PALETTE };

  const tilingAlgorithm =
    dataConfig?.treemapOptions &&
    dataConfig?.treemapOptions.tilingAlgorithm &&
    !isEmpty(dataConfig?.treemapOptions.tilingAlgorithm)
      ? dataConfig?.treemapOptions.tilingAlgorithm[0]
      : 'squarify';

  if (
    isEmpty(data[childField.name]) ||
    isEmpty(data[valueField.name]) ||
    (!isNull(parentField) && isEmpty(data[parentField.name])) ||
    isEqual(childField.name, parentField?.name) ||
    indexOf(NUMERICAL_FIELDS, valueField.type) < 0
  )
    return <EmptyPlaceholder icon={visualizations?.vis?.iconType} />;

  const [treemapData, mergedLayout] = useMemo(() => {
    let labelsArray, parentsArray, valuesArray, colorsArray;

    if (parentField === null) {
      labelsArray = [...data[childField.name]];
      parentsArray = [...Array(labelsArray.length).fill('')];
      valuesArray = [...data[valueField.name]];
      if (colorField.name === MULTI_COLOR_PALETTE) {
        colorsArray = [...Array(data[childField.name].length).fill(colorField.childColor)];
      }
    } else {
      const uniqueParents = uniq(data[parentField.name]);
      labelsArray = [...data[childField.name], ...uniqueParents];
      parentsArray = [...data[parentField.name], ...Array(uniqueParents.length).fill('')];
      valuesArray = [...data[valueField.name], ...Array(uniqueParents.length).fill(0)];
      if (colorField.name === MULTI_COLOR_PALETTE) {
        colorsArray = [
          ...Array(data[childField.name].length).fill(colorField.childColor),
          ...Array(uniqueParents.length).fill(colorField.parentColor),
        ];
      }
    }

    if (colorField.name === SINGLE_COLOR_PALETTE) {
      colorsArray = [...Array(valuesArray.length).fill(colorField.childColor)];
    }

    const markerColors =
      colorField.name === MULTI_COLOR_PALETTE
        ? {
            colors: colorsArray,
          }
        : ![DEFAULT_PALETTE, SINGLE_COLOR_PALETTE].includes(colorField.name)
        ? {
            colorscale: colorField.name,
            colorbar: {
              len: 1,
            },
          }
        : {};

    const colorway = colorField.name === SINGLE_COLOR_PALETTE ? colorsArray : {};

    const mapLayout = {
      ...layout,
      ...(layoutConfig.layout && layoutConfig.layout),
      title: dataConfig?.panelOptions?.title || layoutConfig.layout?.title || '',
      treemapcolorway: colorway,
    };

    const mapData = [
      {
        type: 'treemap',
        labels: labelsArray,
        parents: parentsArray,
        values: valuesArray,
        textinfo: 'label+value+percent parent+percent entry',
        tiling: {
          packing: tilingAlgorithm.value,
        },
        marker: markerColors,
      },
    ];

    return [mapData, mapLayout];
  }, [
    data,
    childField,
    valueField,
    parentField,
    colorField,
    tilingAlgorithm,
    dataConfig,
    layoutConfig,
  ]);

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={treemapData} layout={mergedLayout} config={mergedConfigs} />;
};
