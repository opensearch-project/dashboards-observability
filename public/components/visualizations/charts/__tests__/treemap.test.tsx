/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import {
  LAYOUT_CONFIG,
  TEST_VISUALIZATIONS_DATA,
} from '../../../../../test/event_analytics_constants';
import { TreeMap } from '../maps/treemaps';

describe('Treemap component', () => {
  configure({ adapter: new Adapter() });

  it('Renders treemap component', async () => {
    const wrapper = mount(
      <TreeMap visualizations={TEST_VISUALIZATIONS_DATA} layout={LAYOUT_CONFIG} config={{}} />
    );

    wrapper.update();

    await waitFor(() => {
      expect(
        toJson(wrapper, {
          mode: 'deep',
        })
      ).toMatchSnapshot();
    });
  });
});
