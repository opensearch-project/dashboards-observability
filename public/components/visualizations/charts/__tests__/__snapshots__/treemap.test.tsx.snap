// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treemap component Renders treemap component 1`] = `
<div
  id="explorerPlotComponent"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;
